File: examples/javascript-grpc-listener/src/get_pet.js
/**
 * JavaScript handler for GetPet operation - fetches pet data from petstore API.
 * Demonstrates using third-party libraries via CDN in jackdaw listeners.
 *
 * @param {Object} request - Request containing petId
 * @param {number} request.petId - ID of pet to fetch
 * @returns {Promise<Object>} Pet information from the petstore API
 */
// Using wretch - a tiny wrapper around fetch that works in Deno without extra globals
// This demonstrates using third-party libraries via CDN imports
import wretch from "https://esm.sh/wretch@2.9.0";

export async function handler(request) {
  const petId = request.petId;

  if (!petId) {
    throw new Error("petId is required");
  }

  // Call the petstore API using wretch (third-party HTTP client from CDN)
  // Using https://petstore3.swagger.io which is commonly used in tests
  const url = `https://petstore3.swagger.io/api/v3/pet/${petId}`;

  try {
    // wretch automatically parses JSON responses
    const petData = await wretch(url).get().json();

    // Return the pet data - jackdaw will validate it against the OpenAPI schema
    const result = {
      id: petData.id || 0,
      name: petData.name || "Unknown",
      status: petData.status || "unknown",
    };

    return result;
  } catch (error) {
    // Check for 404 errors
    if (error.status === 404) {
      throw new Error(`Pet ${petId} not found`);
    }
    if (error.status) {
      throw new Error(`Failed to fetch pet: HTTP ${error.status}`);
    }
    // Network or other errors
    throw new Error(`Network error fetching pet: ${error.message}`);
  }
}

File: examples/python-grpc-listener/calculator/get_pet.py
"""
Handler for GetPet RPC - fetches pet data from petstore API.
Demonstrates using third-party dependencies (requests library) in jackdaw gRPC listeners.
"""
import requests
from calculator.types import GetPetRequest, GetPetResponse


def handler(request: GetPetRequest) -> GetPetResponse:
    """
    Fetch pet information from the petstore API.

    Args:
        request: GetPetRequest containing pet_id

    Returns:
        GetPetResponse with pet information from the petstore API
    """
    pet_id = request["pet_id"]

    # Call the petstore API
    # Using https://petstore3.swagger.io which is commonly used in tests
    url = f"https://petstore3.swagger.io/api/v3/pet/{pet_id}"

    try:
        response = requests.get(url, timeout=10)
        response.raise_for_status()

        # Parse the JSON response from petstore
        pet_data = response.json()

        # Return the pet data - jackdaw will validate it against GetPetResponse schema
        result: GetPetResponse = {
            "id": pet_data.get("id", 0),
            "name": pet_data.get("name", "Unknown"),
            "status": pet_data.get("status", "unknown"),
        }

        return result

    except requests.exceptions.HTTPError as e:
        if e.response.status_code == 404:
            raise ValueError(f"Pet {pet_id} not found") from e
        raise ValueError(f"Failed to fetch pet: {e}") from e
    except requests.exceptions.RequestException as e:
        raise ValueError(f"Network error fetching pet: {e}") from e
